### PR TITLE
Add region tags

### PR DESCRIPTION
**Pull request by majosm**
@inducer What would you say to doing something like this to handle volume tagging for geometry-based initialization (and later multi-domain)?

cc @dshtey2

### CODE REVIEW

File: meshmode/mesh/__init__.py [hunk 1]
```diff
@@ -1034,17 +1077,55 @@ def _compute_nodal_adjacency_from_vertices(mesh):
...
+
+    if BTAG_ALL not in boundary_tags:
+        boundary_tags.append(BTAG_ALL)
+    if BTAG_REALLY_ALL not in boundary_tags:
+        boundary_tags.append(BTAG_REALLY_ALL)
+
+    return boundary_tags
+
+
+def index_tags(tags):
```
Comment (inducer) [inline meshmode/mesh/__init__.py:1114]: Does this require that `tags` is narrowed to the non-empty ones first?

File: meshmode/mesh/__init__.py [hunk 2]
```diff
@@ -169,6 +193,11 @@ class MeshElementGroup(Record):
 
         *(dim, nunit_nodes)*
 
+    .. attribute:: regions
```
Comment (inducer) [inline meshmode/mesh/__init__.py:196]: Should this live with the element group or the mesh? (I'm not sure, but I'm kind of leaning mesh.)

File: meshmode/mesh/__init__.py [hunk 3]
```diff
@@ -169,6 +193,11 @@ class MeshElementGroup(Record):
...
+
+        An array *(nelements)*, with the bits of ``regions[i]`` indicating the
+        mesh regions that contain element ``i``.
```
Comment (inducer) [inline meshmode/mesh/__init__.py:199]: If this is allowed to be `None`, document that fact. (here or wherever it ends up)

**Review by inducer — COMMENTED** [review]
Some thoughts below.

File: meshmode/mesh/__init__.py [hunk 4]
```diff
@@ -647,11 +687,6 @@ class Mesh(Record):
         A mapping that maps boundary tag identifiers to their
         corresponding index.
 
-        .. note::
-
-            Elements of :attr:`boundary_tags` that do not cover any
-            part of the boundary will not be keys in this dictionary.
-
```
[the remark this answers is not in the document]
Comment (majosm) [reply meshmode/mesh/__init__.py:654]: As far as I could tell, this wasn't actually true... The code in the mesh constructor didn't check whether the tags were empty.

File: meshmode/mesh/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (majosm) [reply meshmode/mesh/__init__.py:1114]: Should I verbify it? `get_tag_to_index_dict`? Or just `tag_to_index_dict`?

File: meshmode/mesh/__init__.py [hunk 5]
```diff
@@ -1394,6 +1475,26 @@ def check_bc_coverage(mesh, boundary_tags, incomplete_ok=False,
 # }}}
 
 
+# {{{ is_region_tag_empty
+
+def is_region_tag_empty(mesh, region_tag):
+    """Return *True* if the corresponding region tag does not occur as part of
```
[the remark this answers is not in the document]
Comment (majosm) [reply meshmode/mesh/__init__.py:1481]: pydocstyle didn't like this, FWIW. (Not sure if we're following its conventions here too or just in mirgecom.)

[the remark this answers is not in the document]
Comment (inducer) [reply meshmode/mesh/__init__.py:1481]: Nah, I find pydocstyle way too rigid to use everywhere.

File: meshmode/mesh/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (inducer) [reply meshmode/mesh/__init__.py:1114]: :shrug: Not sure. Be consistent with surrounding code? :)

**Comment by majosm** [comment]
Superseded by #320.

**Pull request closed without merging**